File: src/api/plans.ts
import { invoke } from "@tauri-apps/api/core";
import type { PlanSummary, ProductionPlan } from "./types";

/** List all saved plans (lightweight — no target details). */
export async function listPlans(): Promise<PlanSummary[]> {
  return invoke<PlanSummary[]>("list_plans");
}

/** Load a full plan by ID. Returns null if not found. */
export async function getPlan(id: string): Promise<ProductionPlan | null> {
  return invoke<ProductionPlan | null>("get_plan", { id });
}

/** Create or update a plan (matched by plan.id). */
export async function savePlan(plan: ProductionPlan): Promise<void> {
  return invoke("save_plan", { plan });
}

/** Delete a plan by ID. */
export async function deletePlan(id: string): Promise<void> {
  return invoke("delete_plan", { id });
}

File: src-tauri/src/commands/plans.rs
//! Production plan CRUD commands — create, read, update, delete plans stored
//! in the local SQLite database.

use tauri::State;
use crate::db::local::{LocalState, PlanSummary, ProductionPlan};
use super::CommandError;

/// Return lightweight summaries of all saved plans (no target details).
#[tauri::command]
pub fn list_plans(local: State<'_, LocalState>) -> Result<Vec<PlanSummary>, CommandError> {
    local.0.list_plans().map_err(Into::into)
}

/// Load a full plan by ID, including all build targets.
/// Returns `None` if no plan with that ID exists.
#[tauri::command]
pub fn get_plan(
    id: String,
    local: State<'_, LocalState>,
) -> Result<Option<ProductionPlan>, CommandError> {
    local.0.get_plan(&id).map_err(Into::into)
}

/// Create or update a plan (matched by `plan.id`).
#[tauri::command]
pub fn save_plan(
    plan: ProductionPlan,
    local: State<'_, LocalState>,
) -> Result<(), CommandError> {
    local.0.save_plan(&plan).map_err(Into::into)
}

/// Permanently delete a plan by ID.
#[tauri::command]
pub fn delete_plan(
    id: String,
    local: State<'_, LocalState>,
) -> Result<(), CommandError> {
    local.0.delete_plan(&id).map_err(Into::into)
}
